// @azure/playwright/dist/esm/utils/cIInfoProvider.js
var CI_PROVIDERS = /* @__PURE__ */ ((CI_PROVIDERS2) => {
  CI_PROVIDERS2["GITHUB"] = "GITHUB";
  CI_PROVIDERS2["ADO"] = "ADO";
  CI_PROVIDERS2["DEFAULT"] = "DEFAULT";
  return CI_PROVIDERS2;
})(CI_PROVIDERS || {});
class CIInfoProvider {
  static isGitHubActions() {
    return process.env["GITHUB_ACTIONS"] === "true";
  }
  static getCIProvider() {
    if (CIInfoProvider.isGitHubActions()) {
      return "GITHUB" /* GITHUB */;
    } else if (CIInfoProvider.isAzureDevOps()) {
      return "ADO" /* ADO */;
    } else {
      return "DEFAULT" /* DEFAULT */;
    }
  }
  static getCIInfo() {
    const ciProvider = CIInfoProvider.getCIProvider();
    if (ciProvider === "GITHUB" /* GITHUB */) {
      return {
        providerName: "GITHUB" /* GITHUB */,
        branch: this.getGHBranchName() || null,
        author: process.env["GITHUB_ACTOR"] || null,
        commitId: process.env["GITHUB_SHA"] || null,
        revisionUrl: process.env["GITHUB_SERVER_URL"] ? `${process.env["GITHUB_SERVER_URL"]}/${process.env["GITHUB_REPOSITORY"]}/commit/${process.env["GITHUB_SHA"]}` : null
      };
    } else if (ciProvider === "ADO" /* ADO */) {
      return {
        providerName: "ADO" /* ADO */,
        branch: process.env["BUILD_SOURCEBRANCH"] || null,
        author: process.env["BUILD_REQUESTEDFOR"] || null,
        commitId: process.env["BUILD_SOURCEVERSION"] || null,
        revisionUrl: process.env["SYSTEM_TEAMFOUNDATIONCOLLECTIONURI"] ? `${process.env["SYSTEM_TEAMFOUNDATIONCOLLECTIONURI"]}${process.env["SYSTEM_TEAMPROJECT"]}/_git/${process.env["BUILD_REPOSITORY_NAME"]}/commit/${process.env["BUILD_SOURCEVERSION"]}` : null
      };
    } else {
      return {
        providerName: "DEFAULT" /* DEFAULT */,
        branch: process.env["BRANCH"] ?? null,
        author: process.env["AUTHOR"] ?? null,
        commitId: process.env["COMMIT_ID"] ?? null,
        revisionUrl: process.env["REVISION_URL"] ?? null
      };
    }
  }
  static isAzureDevOps() {
    return process.env["AZURE_HTTP_USER_AGENT"] !== void 0 && process.env["TF_BUILD"] !== void 0;
  }
  static getGHBranchName() {
    if (process.env["GITHUB_EVENT_NAME"] === "pull_request" || process.env["GITHUB_EVENT_NAME"] === "pull_request_target") {
      return process.env["GITHUB_HEAD_REF"];
    } else {
      return process.env["GITHUB_REF_NAME"];
    }
  }
}
export {
  CIInfoProvider,
  CI_PROVIDERS
};
